Automattic/wp-calypso
# composite-checkout: Allow calling step isCompleteCallback from inside steps

#### Background

The `@automattic/composite-checkout` package includes the `CheckoutStep` component which can be used to create a series of form steps. Each step has a `isCompleteCallback` prop that can be used to validate a step's content when the user presses the step's "Continue" button.

Sometimes, it's desirable to complete steps programmatically without relying on user input. This can be done by using the function returned by the `useSetStepComplete()` React hook. However, setting the complete status in this way is very heavy-handed, changing the complete status directly and ignoring the `isCompleteCallback` function. This can result in needing to duplicate step validation behavior and other subtle bugs.

#### Changes proposed in this Pull Request

In this PR we modify `useSetStepComplete()` so that its returned function now calls the `isCompleteCallback` prop of the requested step directly as if the user had clicked the "Continue" button. If the callback fails, the step remains incomplete. (Previously it was also possible to use this function to set a step to incomplete, but that functionality has been removed.) In addition, before a step's callback is called, all previous incomplete steps must pass their callbacks as well.

Secondly, as long as we're changing its API, this PR changes the function returned by `useSetStepComplete()` to use the step's `stepId` prop instead of its number. This is a much better argument because steps are numbered automatically but the `stepId` is provided by the developer on the step. If the step numbers were to change, this would allow the function to still target the correct step.

Because the `isCompleteCallback` of the contact/billing step in checkout will now be called by `useSkipToLastStepIfFormComplete()`, and that callback saves the form data to the cart, this PR also removes the redundant `useCachedContactDetailsForCart()`.

#### Testing instructions

**To test basic behavior in demo:**

Run `yarn workspace @automattic/composite-checkout run storybook` and try the prefilled demo content to be sure it works as expected.

**To test calypso checkout:**

Add a product to your cart and visit checkout. Verify that if you have cached contact details, the contact/billing step is automatically completed (and that tax info accurately matches the chosen location). Verify that clicking "Continue" on each step moves to the next step if appropriate.
